File: sdk/MQTT/C++/producer.cpp
/**
 * File: procuder.cpp
 * Description: This is the MQTT producer publishes a certain number
 *     of test messages to a particular topic through MQTT broker.
 *     It will first ask user for the quantity of messages.
 *
 * Author: Eamin Zhang
 * robomq.io (http://www.robomq.io)
 */

#include <stdio.h>
#include <iostream>
#include <mosquitto.h>
#include <exception>
#include <stdlib.h>
#include <unistd.h>

using namespace std;

//The library automatically reconnects to broker

string hst = "hostname";
const char *host = hst.c_str();
int port = 1883;
string vhost = "yourvhost";
string usn = "username";
string vhusn = vhost + ":" + usn;
const char *username = vhusn.c_str();
string pwd = "password";
const char *password = pwd.c_str();
string tpc = "test/any";
const char *topic = tpc.c_str();

/**
 * This is the main method which creates and runs producer instance.
 * @Looping is essential for this MQTT library to work.
 * @Exceptions on connection and publish error.
 */
int main(int argc, char *argv[]) {
	int keepalive = 60;
	bool clean_session = true;
	struct mosquitto *mosq = NULL;

	//create producer and connect to broker
	mosquitto_lib_init();
	mosq = mosquitto_new(NULL, clean_session, NULL);
	mosquitto_username_pw_set(mosq,	username, password);
	if(mosquitto_connect(mosq, host, port, keepalive)) {
		printf("Error: Failed to connect\n");
		return 1;
	}
	//usually start loop right after connecting
	mosquitto_loop_start(mosq); 

	//send certain number of test messages
	int msgNum;
	cout << "Quantity of test messages: ";
	cin >> msgNum;
	char payload[20];
	for (int i = 1; i <= msgNum; i++) {
		sprintf(payload, "test msg %d", i);
		try {
			mosquitto_publish(mosq, NULL, topic, 20, payload, 1, false);
		} catch(exception& e) {
			printf("Error: Failed to publish message\n%s\n", e.what());
			return 1;
		}
		sleep(1);
	}

	//stop producer
	mosquitto_loop_stop(mosq, true); 
	mosquitto_disconnect(mosq);
	mosquitto_destroy(mosq);
	mosquitto_lib_cleanup();
	return 0;
}

File: sdk/MQTT/Java/Producer.java
/**
 * File: Procuder.java
 * Description: This is the MQTT producer publishes a certain number
 *     of test messages to a particular topic through MQTT broker.
 *     It will first ask user for the quantity of messages.
 *
 * Author: Eamin Zhang
 * robomq.io (http://www.robomq.io)
 */

import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
import java.util.Scanner;

public class Producer {

	private MqttClient client;
	private String server = "hostname";
	private String port = "1883";
	private String broker = "tcp://" + server + ":" + port;
	private String vhost = "yourvhost";
	private String username = "username";
	private String password = "password";
	private String topic = "test/any";
	private String clientId = MqttClient.generateClientId();
	private MemoryPersistence persistence = new MemoryPersistence();

	private void produce() {
		try {
			client = new MqttClient(broker, clientId, persistence);
			MqttConnectOptions connOpts = new MqttConnectOptions();
			connOpts.setUserName(vhost + ":" + username);
			connOpts.setPassword(password.toCharArray());
			connOpts.setKeepAliveInterval(60);
			connOpts.setCleanSession(true);
			client.connect(connOpts);
			System.out.print("Quantity of test messages: ");
			Scanner scanner = new Scanner(System.in);
			int msgNum = scanner.nextInt();
			for (int i = 0; i < msgNum; i ++) {
				MqttMessage message = new MqttMessage(("test msg " + Integer.toString(i + 1)).getBytes());
				message.setQos(1);
				message.setRetained(false);
				client.publish(topic, message);
				try {
					Thread.sleep(1000);
				} catch (Exception e) {}
			}
			client.disconnect();
		} catch(MqttException me) {
			System.out.println(me);
			System.exit(-1);			
		}
	}

	public static void main(String[] args) {
		Producer p = new Producer();
		p.produce();
	}
}
